### DISCUSSION TITLE
Handling errors in actions without triggering the loaders and loosing the content of the inputs

### DISCUSSION THREAD

**Discussion by lsdebianchi** [remix-run/remix#4516] [Q&A] [answered]
### TLDR; Is there a way to ask remix to not re-trigger the loaders after an action?

CONTEXT

I am developing a sort of CMS so my front-end is quite interactive and for this reason all my input are controlled and they get their initial values from a remix loader which fetches them from a database.

So let's imagine I have an inout field in my UI with the initial value (coming from the loader) of:

```
"my title"
```

Now the user changes it to:

```
"My ßeautiful Title"
```

and submits the form through an action.

The action submits the data to the main server and db which complains that "ß" is an invalid character.
So I return my response with `status: "error"` and some custom data to show to the user a nice notification about the failed validation.

My problem is that after such "failed" action, remix re-triggers the loaders, thus re-fetching the initial values from the db, re-rendering the page (react) and re-populating the inout fields.

**The final result is that the user loses his/her progress every time that a submission fails for validation.**

QUESTION

- How people normally handle this?
- Is there a way for a remix action to not re-trigger the loaders?
- Obviously I cannot use the `CatchBoundary`  & `ErrorBoundary` because I need to keep my UI with the user filled inputs.
- I could try to protect the inputs from refreshing with some react magic but it feels like the bad pattern, as I think that they should always refresh when a data mutation actually succede.

**Comment by sergiodxa** [comment, marked as answer]
You need to export a unstable_shouldReload function from the route with the loaders you want to prevent the reload.

**Reply by lsdebianchi** [reply]
Thanks, very interesting! I must have missed that.
I am having some difficulty making it take effect but it's a step in the good direction.

**Comment by lsdebianchi** [comment]
This is very related:
https://github.com/remix-run/remix/issues/3069

**Comment by christemple** [comment]
@lsdebianchi In the event of a form error in the `action` function, you can return the current field values alongside the field errors so that the component can render the form with the same values you just submitted

The [Jokes app tutorial (Mutations)](https://remix.run/docs/en/v1/tutorials/jokes#mutations) has a good example of this pattern with `fields` and `fieldErrors` being returned in the event of an invalid form (bad request), you can then pre-fill your form back to exactly how it was submitted using `defaultValue`

(scroll past the screenshots and look at the next codeblock in the Mutations section for the example)
I've literally just finished the tutorial and completely new to Remix, so somebody correct me if I'm wrong!

**Reply by lsdebianchi** [reply]
Good point! I see.
It passed through my head as well that perhaps the action could return the user's edits.
It offers a lot of control but it's also quite an involved pattern.

Yet you still need to deactivate the loaders or they will refetch irrelevant data after the failed actions anyway (which is a wasteful request).
